From a2fc16e39b2f589c9d179eaea1f4cf2a13d30fb9 Mon Sep 17 00:00:00 2001
From: Xiaolin He <xiaolin.he@nxp.com>
Date: Tue, 6 Aug 2019 13:34:18 +0800
Subject: [PATCH 659/741] tsn-switch: add support for no flowmeter assigned for
 stream filter instance case.

commit c15d31d79d5bfb2c41a9983499b16c57a5105b6b from
http://source.codeaurora.org/external/qoriq/qoriq-components/linux

Use the max available flowmeter index as policer index.

Signed-off-by: Xiaolin He <xiaolin.he@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 drivers/net/ethernet/mscc/felix_tsn.c | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/drivers/net/ethernet/mscc/felix_tsn.c b/drivers/net/ethernet/mscc/felix_tsn.c
index d733e699ddf4..5ea210c91067 100644
--- a/drivers/net/ethernet/mscc/felix_tsn.c
+++ b/drivers/net/ethernet/mscc/felix_tsn.c
@@ -902,10 +902,16 @@ int felix_qci_sfi_set(struct net_device *ndev, u32 index, bool enable,
 	struct ocelot *ocelot = port->ocelot;
 	int igr_prio = sfi->priority_spec;
 	u16 sgid  = sfi->stream_gate_instance_id;
-	u16 pol_idx = sfi->stream_filter.flow_meter_instance_id;
+	u16 pol_idx;
+	int fmid = sfi->stream_filter.flow_meter_instance_id;
 	u16 max_sdu_len = sfi->stream_filter.maximum_sdu_size;
 	int sfid = index;
 
+	if (fmid == -1)
+		pol_idx = capa.psfp_fmi_max;
+	else
+		pol_idx = (u16)fmid;
+
 	if (sfid >= capa.num_psfp_sfid) {
 		netdev_info(ndev, "Invalid index %u, maximum:%u\n",
 			    sfid, capa.num_psfp_sfid);
@@ -925,7 +931,7 @@ int felix_qci_sfi_set(struct net_device *ndev, u32 index, bool enable,
 
 	ocelot_write(ocelot, ANA_TABLES_SFIDTIDX_SGID_VALID |
 		     ANA_TABLES_SFIDTIDX_SGID(sgid) |
-		     ANA_TABLES_SFIDTIDX_POL_ENA |
+		     ((fmid != -1) ? ANA_TABLES_SFIDTIDX_POL_ENA : 0) |
 		     ANA_TABLES_SFIDTIDX_POL_IDX(pol_idx) |
 		     ANA_TABLES_SFIDTIDX_SFID_INDEX(sfid),
 		     ANA_TABLES_SFIDTIDX);
-- 
2.17.1

